h/services/bulk_executor/executors.py

```diff
@@ -0,0 +1,90 @@
+from sqlalchemy.orm.exc import NoResultFound
+
+from h.h_api.bulk_api.executor import Executor
+from h.h_api.enums import CommandType, DataType
+from h.h_api.exceptions import InvalidDeclarationError, UnsupportedOperationError
+from h.models import User
+from h.services.bulk_executor.actions import (
+    GroupMembershipCreateAction,
+    GroupUpsertAction,
+    UserUpserAction,
+)
+
+
+class DBExecutor(Executor):
+    def __init__(self, db):
+        self.db = db
+        self.effective_user_id = None
+
+    def configure(self, config):
+        try:
+            user = (
+                self.db.query(User).filter(User.userid == config.effective_user).one()
+            )
+
+        except NoResultFound:
+            raise InvalidDeclarationError(
+                f"No user found for effective user: '{config.effective_user}'"
+            )
+
+        self.effective_user_id = user.id
+
+    def execute_batch(self, command_type, data_type, default_config, batch):
+        default_config["effective_user_id"] = self.effective_user_id
+
+        handler = self._get_handler(command_type, data_type)
+        return handler(self.db).execute(batch, **default_config)
+
+    def _get_handler(self, command_type, data_type):
+        # This would be nicer as a class attribute, but it makes testing a pain
+        handler = {
+            (CommandType.UPSERT, DataType.USER): UserUpserAction,
+            (CommandType.UPSERT, DataType.GROUP): GroupUpsertAction,
+            (
+                CommandType.CREATE,
+                DataType.GROUP_MEMBERSHIP,
+            ): GroupMembershipCreateAction,
+        }.get((command_type, data_type), None)
+
+        if handler is None:
+            raise UnsupportedOperationError(
+                f"No implementation for {command_type.value} {data_type.value}"
+            )
+
+        return handler
+
+
+class AuthorityCheckingExecutor(DBExecutor):
```

Comment:
I think it'd be nicer to use compostion over inheritance.

This class isn't really an authority-checking executor. It's an authority-checking-and-also-DB-writing-executor.

One possible approach might be:

```python
class ComboExecutor:
    def __init__(self, executors):
        self._executors = executors

    def configure(self, *args, **kwargs):
        for executor in self._executors:
            executor.configure(*args, **kwargs)

    def execute_batch(self, *args, **kwargs):
        for executor in self._executors:
            executor.execute_batch(*args, **kwargs)

def bulk_executor_service_factory(_, request):
    return ComboExecutor([AuthorityCheckingExecutor(), DBExecutor(request.db)])
```

`AuthorityCheckingExecutor` then wouldn't need the DB, inheritance or super calls.

Reply:
`AuthorityCheckingExecutor` code should be simpler to understand as you no longer need to think about `DBExecutor` to understand it. With inheritance `AuthorityCheckingExecutor` has a `db` param that's actually `DBExecutor`'s param, for example, and it knows about `DBExecutor.__init__()`'s signature.

This would also mean that `AuthorityCheckingExecutor` could be unit-tested without also executing `DBExecutor`'s code and touching the DB (with inheritance, you can't mock out the base class so you can't test `AuthorityCheckingExecutor` without `DBExecutor`).

Similarly `ComboExecutor` can be both understood and unit-tested without having to know anything about the other two executors.

It's more flexible too, although we don't need that flexibility yet. Currently the only sequence possible is `AuthorityCheckingExecutor` -> `DBExecutor`, as that's hardcoded in the inheritance. But `ComboExecutor` can combine any set of executors in any order.

Reply:
The other thing I think I might do, is just merge the two together. The whole hot-swapping executors thing strikes me as YAGNI. I don't think we'll ever have another behavior.

Reply:
It also just occurred to me (facepalm) that if there's only one authority, we can enforce it in the schema and avoid all of this?

It's not as "flexible" but why worry about that? This would have the benefit of being more specific about what you got wrong too.

Reply:
Gah... me from the past/future. The reason this was decided against is it's not the libraries job to police authentication style issues, that's should reside in H.

Reply:
Ok, so I've come down on the side of merging the two which means there is no more inheritance problem to solve. The classes' job is to prep for execution and do it.

It's smaller than the previous two and probably doesn't merit chopping up much.